File: android/src/main/kotlin/dev/masel/synced_bloc/SyncMaster.kt
package dev.masel.synced_bloc

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

internal class SyncMaster(
    binaryMessenger: BinaryMessenger,
    val masterId: String
) {
    private val channel = MethodChannel(binaryMessenger, "dev.masel.synced_bloc.master.$masterId")

    fun getState(onSuccess: (res: Any?) -> Unit, onError: () -> Unit) {
        channel.invokeMethod(
            "getState",
            null,
            object : MethodChannel.Result {
                override fun success(res: Any?) = onSuccess(res)

                override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) =
                    onError()

                override fun notImplemented() = onError()
            })
    }

    fun addEvent(event: String) =
        channel.invokeMethod("addEvent", event)
}

File: ios/Classes/SyncMaster.swift
import Flutter

internal class SyncMaster {
    private let channel: FlutterMethodChannel
    let masterId: String

    init(binaryMessenger: FlutterBinaryMessenger, masterId: String) {
        self.channel = FlutterMethodChannel(name: "dev.masel.synced_bloc.master.\(masterId)", binaryMessenger: binaryMessenger)
        self.masterId = masterId
    }

    func getState(onSuccess: @escaping (Any?) -> Void, onError: @escaping () -> Void) {
        channel.invokeMethod("getState", arguments: nil) { result in
            if let res = result {
                onSuccess(res)
            } else {
                onError()
            }
        }
    }

    func addEvent(event: String) {
        channel.invokeMethod("addEvent", arguments: event)
    }
}
